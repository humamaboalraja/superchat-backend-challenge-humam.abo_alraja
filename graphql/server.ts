import { ApolloServer } from 'apollo-server-express';
import express from 'express';
import { createServer } from 'http';
import { createContext } from './context';
import { schema } from './schema';

const { PORT = 5000 } = process.env;

const app = express();

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: process.env.STAGE !== 'production',
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
