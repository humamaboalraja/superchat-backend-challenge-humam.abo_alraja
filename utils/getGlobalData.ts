import { prisma } from '../singletons/prisma';

export default async function getGlobalData() {
  const messages = await prisma.message.findMany({});
  const contacts = await prisma.contact.findMany({});
  if (!messages || !contacts) throw new Error('Unable to set up global data');
  return { messages, contacts };
}
