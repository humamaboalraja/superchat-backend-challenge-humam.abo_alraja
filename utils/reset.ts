import { prisma } from '../singletons/prisma';

export default async function reset() {
  try {
    const models: string[] = ['Message', 'Contact'];
    for (const model in models) {
      await prisma.$executeRaw(`DELETE FROM "${models[model]}";`);
    }
  } catch (e) {
    throw new Error(`${e}`);
  }
}
