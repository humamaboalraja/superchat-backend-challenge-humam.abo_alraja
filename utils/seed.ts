import cryptoRandomString from 'crypto-random-string';
import { prisma } from '../singletons/prisma';

export default async function seed() {
  try {
    const contact1 = await prisma.contact.create({
      data: {
        name: 'Humam Abo Alraja',
        email: 'humam@company.com',
      },
    });
    const contact2 = await prisma.contact.create({
      data: {
        name: 'John',
        email: 'john@company.com',
      },
    });

    const contact3 = await prisma.contact.create({
      data: {
        name: 'Emma Mockassin',
        email: 'emma@company.com',
      },
    });

    const contact4 = await prisma.contact.create({
      data: {
        name: 'Alex Smith',
        email: 'alex@company.com',
      },
    });

    if (!contact1 || !contact2 || !contact3 || !contact4)
      throw new Error('no contacts');

    const one = await prisma.message.create({
      data: {
        messageContent: 'Hey Emma, How is it going?!',
        sender: {
          connect: {
            id: contact1.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });

    const two = await prisma.message.create({
      data: {
        messageContent: "Hi Humam, it's going great! How about you?",
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact1.id,
          },
        },
      },
    });

    const three = await prisma.message.create({
      data: {
        messageContent:
          'Hey Alex, do you mind sending me the pitch ASAP? Thanks in advance!',
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact4.id,
          },
        },
      },
    });

    const four = await prisma.message.create({
      data: {
        messageContent: 'Hi Emma, Definitely!',
        sender: {
          connect: {
            id: contact4.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });
  } catch (e) {
    throw new Error(`${e}`);
  }
}
