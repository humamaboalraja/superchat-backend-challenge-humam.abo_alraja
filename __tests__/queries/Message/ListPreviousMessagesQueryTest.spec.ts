import getGlobalData from '../../../utils/database/getGlobalData';
import reset from '../../../utils/database/reset';
import seed from '../../../utils/database/seed';
import server from '../../server';

beforeAll(async () => {
  await reset();
  await seed();
  global.testData = await getGlobalData();
});

const LIST_PREVIOUS_MESSAGES_QUERY = `
query ListPreviousMessages($senderId: String, $recieverId: String) {
   listPreviousMessages(one: $senderId, two: $recieverId) {
     senderId
     recieverId
     messageContent
   }
 }
  `;

describe('Test find contact query', () => {
  it('should list previous messages between two contacts', async () => {
    const res = await server.executeOperation({
      query: LIST_PREVIOUS_MESSAGES_QUERY,
      variables: {
        one: global.testData.contacts[1].id,
        two: global.testData.contacts[2].id,
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('should not find previous messages between two contacts if no arguments are passed in', async () => {
    const res = await server.executeOperation({
      query: LIST_PREVIOUS_MESSAGES_QUERY,
      variables: {
        one: '',
        two: '',
      },
    });
    expect(res).toMatchSnapshot();
  });
});

export {};
