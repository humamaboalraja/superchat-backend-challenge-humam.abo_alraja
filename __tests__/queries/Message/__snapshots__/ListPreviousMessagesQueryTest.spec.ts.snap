// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test find contact query should list previous messages between two contacts 1`] = `
Object {
  "data": Object {
    "listPreviousMessages": Array [
      Object {
        "messageContent": "Hey Emma, How is it going?!",
        "recieverId": "ckwdfh74e0014t5hfv8sf5lfe",
        "senderId": "ckwdfh71d0000t5hfrman1lvo",
      },
      Object {
        "messageContent": "Hi Humam, it's going great! How about you?",
        "recieverId": "ckwdfh71d0000t5hfrman1lvo",
        "senderId": "ckwdfh74e0014t5hfv8sf5lfe",
      },
      Object {
        "messageContent": "Hey Alex, do you mind sending me the pitch ASAP? Thanks in advance!",
        "recieverId": "ckwdfh76r0021t5hfxy22j582",
        "senderId": "ckwdfh74e0014t5hfv8sf5lfe",
      },
      Object {
        "messageContent": "Hi Emma, Definitely!",
        "recieverId": "ckwdfh74e0014t5hfv8sf5lfe",
        "senderId": "ckwdfh76r0021t5hfxy22j582",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
